### DISCUSSION TITLE
Underscores in figure labels when rendering via LaTeX to pdf?

### DISCUSSION THREAD

**Discussion by nonsciencemark** [quarto-dev/quarto-cli#8630] [Q&A]
### Description

I'm not sure what changed on my system, but I suppose some package updated and now including figures such as:

```
![This is a figure caption.](this/is/a/figure.png){#fig-this_didnt_work}
```

caused failure to compille, while

```
![This is a figure caption.](this/is/a/figure.png){#fig-this-DID-work}
```

is fine. Is this normal? The document used to compile fine and then stopped and, because I'm completely lost with LaTeX in general, I'm not sure why.

---

The error message in my actual script was: 

```
compilation failed- error
Missing \endcsname inserted.
<to be read again>
                   \protect
l.18 ...ifferent treatments}{figure.caption.11}{}}
```

Quarto version: 1.4.549
System: Windows 10

**Comment by mcanouil** [comment]
LaTeX does not support underscores in labels, see the callout below the image <https://quarto.org/docs/authoring/cross-references.html>.

- https://github.com/quarto-dev/quarto-cli/issues/8374
- https://github.com/quarto-dev/quarto-cli/issues/7307

**Reply by nonsciencemark** [reply]
I see. Was it just a lucky error that it _was_ working before, then? Or is this a recent change?

**Reply by cscheid** [reply]
I can't understand why this would have ever worked. If you can produce a document with underscores and references that renders to LaTeX and works in 1.3 but doesn't in 1.4 we would be happy to take a look at it.

**Reply by nonsciencemark** [reply]
I can't re-render through Quarto but I attach a tex file (saved as just .txt so that I can actually upload it) that was _produced by Quarto_ (I've edited to remove unneeded stuff, but left all the LaTeX preamble in case it helps) that still successfully compiles with underscores in the figure labels (pdf attached). 

In the tex, the figure label is defined on line 229 as `\caption{\label{fig-cilia_td_general}caption.}` (in Quarto it was `![caption.](figures/cilia/cilia_td_general.pdf){#fig-cilia_td_general width="4in"})`. If you need more let me know - it's hard to recreate the issue but I can work on it if it helps 

[demo_tex.txt](https://github.com/quarto-dev/quarto-cli/files/14211986/demo_tex.txt) (this is the tex file with file extension changed to allow upload)
[demo_tex.pdf](https://github.com/quarto-dev/quarto-cli/files/14211965/demo_tex.pdf)

**Reply by mcanouil** [reply]
Plain text can just copy/pasted inside code blocks, no need to upload anything.

Since here you are saying you were able to produce LaTeX/PDF from Quarto 1.3.450, we need to replicate that.
So we need the original Quarto document (or Jupyter notebook).

**Reply by cderv** [reply]
Thanks for pushing us to re-look into that. This example is indeed working in 1.3.450
````markdown
---
format: pdf
keep-tex: true
---

![This is a figure caption.](surus.png){#fig-this_didnt_work}

See @fig-this_didnt_work.

````

File is from here: [surus.png](https://github.com/quarto-dev/quarto-web/blob/main/docs/authoring/_figure-examples/surus.png)

But it does not work in 1.4. Difference in tex file is 
````diff
-\caption{\label{fig-this_didnt_work}This is a figure caption.}
+\caption{\label{fig-this\_didnt\_work}This is a figure caption.}
````

So it seems we do escape `_` to `\_` because usually it needs to be in LaTeX, but inside `\label` is does not seem to be required.

I can't say if this our change, or a Pandoc change... we can look if this specific case could be change back, but usually we don't expect underscore in label and this could cause other issues. So I would go that it was working by chance previously...

**Reply by cscheid** [reply]
It's on us. Pandoc's behavior is that is _always_ escapes this. 3.1.1 (1.3's pandoc):

```
$ quarto-1.3 pandoc -f markdown -t latex latex-escape.qmd
\begin{figure}
\hypertarget{fig-this_didnt_work}{%
\centering
\includegraphics{surus.png}
\caption{This is a figure caption.}\label{fig-this_didnt_work}
}
\end{figure}

See @fig-this\_didnt\_work.

$ quarto-1.4 pandoc -f markdown -t latex latex-escape.qmd
\begin{figure}
\centering
\includegraphics{surus.png}
\caption{This is a figure caption.}\label{fig-this_didnt_work}
\end{figure}

See @fig-this\_didnt\_work.
```

But the `quarto render` output is wrong in 1.4:

```
$ tail q14.tex

\caption{\label{fig-this\_didnt\_work}This is a figure caption.}

\end{figure}%

See Figure~\ref{fig-this_didnt_work}.



\end{document}
$ tail q13.tex

\caption{\label{fig-this_didnt_work}This is a figure caption.}

\end{figure}

See Figure~\ref{fig-this_didnt_work}.



\end{document}
```

This is a regression. Let's fix it.

**Reply by nonsciencemark** [reply]
Sorry @mcanouil - I didn't get back to you with the Quarto document in a timely fashion. It seems you've found a root cause of it though, cool. And going forward, no underscores, got it !
